src/dex-v2.ts

```diff
@@ -1003,4 +1000,111 @@ export class DexV2 {
     }
     return lucidTx.complete();
   }
+
+  async cancelExpiredOrders({
+    orderUtxos,
+    currentSlot,
+    extraDatumMap,
+  }: CancelExpiredOrderOptions): Promise<TxComplete> {
+    const refScript = await this.lucid.utxosByOutRef([
+      DexV2Constant.DEPLOYED_SCRIPTS[this.networkId].order,
+      DexV2Constant.DEPLOYED_SCRIPTS[this.networkId].expiredOrderCancellation,
+    ]);
+    const currentTime = this.lucid.utils.slotToUnixTime(currentSlot);
+    invariant(
+      refScript.length === 2,
+      "cannot find deployed script for V2 Order or Expired Order Cancellation"
+    );
+    const sortedOrderUtxos = [...orderUtxos].sort(compareUtxo);
+
+    const lucidTx = this.lucid.newTx().readFrom(refScript);
+    lucidTx.collectFrom(
+      sortedOrderUtxos,
+      Data.to(new Constr(OrderV2.Redeemer.CANCEL_EXPIRED_ORDER_BY_ANYONE, []))
+    );
+    for (const orderUtxo of sortedOrderUtxos) {
+      const orderAddr = orderUtxo.address;
+      const orderScriptPaymentCred =
+        this.lucid.utils.getAddressDetails(orderAddr).paymentCredential;
+      invariant(
+        orderScriptPaymentCred?.type === "Script" &&
+          orderScriptPaymentCred.hash ===
+            DexV2Constant.CONFIG[this.networkId].orderScriptHash,
+        `Utxo is not belonged Minswap's order address, utxo: ${orderUtxo.txHash}`
+      );
+      let datum: OrderV2.Datum;
+      if (orderUtxo.datum) {
+        const rawDatum = orderUtxo.datum;
+        datum = OrderV2.Datum.fromPlutusData(
+          this.networkId,
+          Data.from(rawDatum)
+        );
+      } else if (orderUtxo.datumHash) {
+        const rawDatum = await this.lucid.datumOf(orderUtxo);
+        datum = OrderV2.Datum.fromPlutusData(
+          this.networkId,
+          rawDatum as Constr<Data>
+        );
+      } else {
+        throw new Error(
+          "Utxo without Datum Hash or Inline Datum can not be spent"
+        );
+      }
+      const expiryOptions = datum.expiredOptions;
+      invariant(expiryOptions !== undefined, "Order must have expiry options");
+      invariant(
+        expiryOptions.maxCancellationTip >= DexV2Constant.DEFAULT_CANCEL_TIPS,
+        "Cancel tip is too low"
+      );
+      invariant(
+        expiryOptions.expiredTime < BigInt(currentTime),
+        "Order is not expired"
+      );
+      const refundDatum = datum.refundReceiverDatum;
+      const outAssets = { ...orderUtxo.assets };
+      outAssets["lovelace"] -= expiryOptions.maxCancellationTip;
+      switch (refundDatum.type) {
+        case OrderV2.ExtraDatumType.NO_DATUM: {
+          lucidTx.payToAddress(datum.refundReceiver, outAssets);
+          break;
+        }
+        case OrderV2.ExtraDatumType.DATUM_HASH: {
+          invariant(
```

Comment:
If the datum is found, then attach the datum to the transaction. Otherwise, giving the hash instead of